src/txreconciliation.h

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_TXRECONCILIATION_H
+#define BITCOIN_TXRECONCILIATION_H
+
+#include <net.h>
+#include <sync.h>
+
+#include <memory>
+#include <tuple>
+
+/**
+ * Transaction reconciliation is a way for nodes to efficiently announce transactions.
+ * This object keeps track of all reconciliation-related communications with the peers.
+ * The high-level protocol is:
+ * 0. Reconciliation protocol handshake.
+ * 1. Once we receive a new transaction, add it to the set instead of announcing immediately
+ * 2. When the time comes, a reconciliation initiator requests a sketch from the peer, where a sketch
+ *    is a compressed representation of their set
+ * 3. Once the initiator received a sketch from the peer, the initiator computes a local sketch,
+ *    and combines the two skethes to find the difference in *sets*.
+ * 4. Now the initiator knows full symmetrical difference and can request what the initiator is
+ *    missing and announce to the peer what the peer is missing. For the former, an extra round is
+ *    required because the initiator knows only short IDs of those transactions.
+ * 5. Sometimes reconciliation fails if the difference is larger than the parties estimated,
+ *    then there is one sketch extension round, in which the initiator requests for extra data.
+ * 6. If extension succeeds, go to step 4.
+ * 7. If extension fails, the initiator notifies the peer
+ */
```

Comment:
Thanks for the clear write-up, may a suggest to insert a stable link to the paper ?